### PR TITLE
chainntnfs+lnrpc/chainrpc: add ChainNotifier RPC subserver

### PR DESCRIPTION
**Pull request by wpaulino**
In this PR, we introduce a new RPC subserver for lnd's ChainNotifier as defined in `chainnotifier.proto`. This new subserver has its own macaroon state and gives users a simplified
interface for a gRPC service that is able to dispatch notifications for blocks, transaction/script confirmations, and output/script spends.

In order to satisfy some of the requirements of the service, some changes were necessary to lnd's ChainNotifier. These changes include:

* Dispatching confirmation and spend notifications for scripts (achieved by #2291).
* Dispatching a block notification for the current tip upon registration.
* Adding a Done channel to signal when notifications are no longer under the risk of being reorged out of the chain.

In order to enable this service, lnd must be built with the `chainrpc` build tag.

Depends on #2291.

### CODE REVIEW

**Comment by joostjager** [comment]
It doesn't build:
```
joost@joost-ThinkPad-T440s $ make  tags="signrpc walletrpc chainrpc"
 Building debug lnd and lncli.
GO111MODULE=on go build -v -tags="dev signrpc walletrpc chainrpc" -o lnd-debug -ldflags "-X github.com/lightningnetwork/lnd/build.Commit=v0.5.1-beta-172-g45fad52f1e7ddf027c10c926fc6065198945c0a3" github.com/lightningnetwork/lnd
github.com/lightningnetwork/lnd/lnrpc/chainrpc
# github.com/lightningnetwork/lnd/lnrpc/chainrpc
lnrpc/chainrpc/chainnotifer_server.go:209:2: undefined: confEvent
lnrpc/chainrpc/chainnotifer_server.go:209:13: undefined: err
lnrpc/chainrpc/chainnotifer_server.go:212:5: undefined: err
lnrpc/chainrpc/chainnotifer_server.go:213:10: undefined: err
lnrpc/chainrpc/chainnotifer_server.go:215:8: undefined: confEvent
lnrpc/chainrpc/chainnotifer_server.go:224:20: select case must be receive, send or assign recv
lnrpc/chainrpc/chainnotifer_server.go:224:25: undefined: confEvent
lnrpc/chainrpc/chainnotifer_server.go:251:14: select case must be receive, send or assign recv
lnrpc/chainrpc/chainnotifer_server.go:251:19: undefined: confEvent
lnrpc/chainrpc/chainnotifer_server.go:266:19: undefined: confEvent
lnrpc/chainrpc/chainnotifer_server.go:266:19: too many errors
Makefile:99: recipe for target 'build' failed
make: *** [build] Error 2
```

**Comment by joostjager** [comment]
Reminder: immediately send current block after `RegisterBlockEpochNtfn`

**Comment by joostjager** [comment]
Ntfr log not enabled in main `log.go`.

**Comment by joostjager** [comment]
If there is so much duplication, can't more be extracted into reusable methods first and build the new functionality on top of that?

`dupl -t 80`

```
found 2 clones:
  bitcoindnotify/bitcoind.go:332,361
  btcdnotify/btcd.go:381,410
found 2 clones:
  bitcoindnotify/bitcoind.go:228,261
  btcdnotify/btcd.go:296,329
found 3 clones:
  bitcoindnotify/bitcoind.go:199,225
  btcdnotify/btcd.go:268,293
  neutrinonotify/neutrino.go:282,307
found 2 clones:
  txnotifier.go:499,527
  txnotifier.go:821,849
found 2 clones:
  bitcoindnotify/bitcoind.go:478,637
  btcdnotify/btcd.go:529,689
found 2 clones:
  height_hint_cache.go:119,233
  height_hint_cache.go:236,351
found 3 clones:
  bitcoindnotify/bitcoind.go:1027,1106
  btcdnotify/btcd.go:1041,1122
  neutrinonotify/neutrino.go:946,1025
found 2 clones:
  bitcoindnotify/bitcoind_test.go:32,49
  btcdnotify/btcd_test.go:30,47
found 2 clones:
  bitcoindnotify/bitcoind.go:282,318
  btcdnotify/btcd.go:331,367
found 2 clones:
  bitcoindnotify/bitcoind.go:735,749
  btcdnotify/btcd.go:796,810

Found total 10 clone groups.
```
This is just automated output, if the code is slightly different it isn't reported by `dupl`.

**Comment by joostjager** [comment]
Tried to build several commits with `git bisect` and manually too. Seems that many of them aren't building.

File: chainntnfs/interface.go [hunk 1]
```diff
@@ -199,6 +199,12 @@ type ConfirmationEvent struct {
 	// NOTE: This channel must be buffered.
 	NegativeConf chan int32
 
+	// Done is a channel that gets sent upon once the confirmation request
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/interface.go]: Guessing this comment can be ignored as you were still getting through the diff? The `Done` channel is used to indicate that the stream can be closed from the server side.

File: chainntnfs/height_hint_cache.go [hunk 2]
```diff
@@ -136,14 +137,20 @@ func (c *HeightHintCache) CommitSpendHint(height uint32, ops ...wire.OutPoint) e
...
+					&outpoint, spendRequest.OutPoint,
+				)
+				if err != nil {
+					return err
+				}
+				spendRequestKey = outpoint.Bytes()
 			}
 
-			err = spendHints.Put(outpoint.Bytes(), hint.Bytes())
+			err := spendHints.Put(spendRequestKey, hint.Bytes())
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/height_hint_cache.go]: If this is undesirable, we could just add additional buckets for them. Let me know if we should proceed with that or stay with the current design.

File: chainntnfs/txnotifier.go [hunk 3]
```diff
@@ -10,6 +10,20 @@ import (
 	"github.com/btcsuite/btcutil"
 )
 
+var (
+	// ZeroHash is the value that should be used as the txid when
+	// registering for the confirmation of a script on-chain. This allows
+	// the notifier to match _and_ dispatch upon the inclusion of the script
+	// on-chain, rather than the txid..
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/txnotifier.go]: Fixed.

File: chainntnfs/txnotifier.go [hunk 4]
```diff
@@ -94,6 +108,46 @@ func newSpendNtfnSet() *spendNtfnSet {
...
+	// PkScript is the public key script of an outpoint created in this
+	// transaction.
+	PkScript txscript.PkScript
+}
+
+// NewConfRequest creates a request for a confirmation notification of either a
+// txid or output script.
+func NewConfRequest(txid *chainhash.Hash, pkScript []byte) (ConfRequest, error) {
+	var r ConfRequest
+	outputScript, err := txscript.ParsePkScript(pkScript)
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/txnotifier.go]: Done.

File: chainntnfs/txnotifier.go [hunk 5]
```diff
@@ -94,6 +108,46 @@ func newSpendNtfnSet() *spendNtfnSet {
...
+		r.TxID = *txid
+	}
+	r.PkScript = outputScript
+
+	return r, nil
+}
+
+// String returns the string representation of the ConfRequest.
+func (c ConfRequest) String() string {
+	if c.TxID != ZeroHash {
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/txnotifier.go]: Any reason to do this? Would made logs much more verbose, especially for longer scripts.

File: chainntnfs/height_hint_cache.go [hunk 6]
```diff
@@ -136,14 +137,20 @@ func (c *HeightHintCache) CommitSpendHint(height uint32, ops ...wire.OutPoint) e
...
 
-		for _, op := range ops {
-			var outpoint bytes.Buffer
-			err := channeldb.WriteElement(&outpoint, op)
-			if err != nil {
-				return err
+		for _, spendRequest := range spendRequests {
+			spendRequestKey := spendRequest.PkScript.Script()
+			if spendRequest.OutPoint != ZeroOutPoint {
+				var outpoint bytes.Buffer
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/height_hint_cache.go]: Fixed.

File: chainntnfs/txnotifier.go [hunk 7]
```diff
@@ -1032,117 +1034,204 @@ func (n *TxNotifier) ConnectTip(blockHash *chainhash.Hash, blockHeight uint32,
...
+
+	// In order to determine if this transaction is relevant to the
+	// notifier, we'll check its inputs for any outstanding spend
+	// requests.
+	txHash := tx.Hash()
+	if onSpend != nil {
+		// constructDetails is a helper closure that will construct the
+		// spend details of a request and hand them off to the onSpend
+		// callback.
+		constructDetails := func(spendRequest SpendRequest,
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/txnotifier.go]: Fixed.

File: chainntnfs/txnotifier.go [hunk 8]
```diff
@@ -1032,117 +1034,204 @@ func (n *TxNotifier) ConnectTip(blockHash *chainhash.Hash, blockHeight uint32,
...
+}
 
-	// Finally, now that we've determined which transactions were confirmed
-	// and which outpoints were spent within the new block, we can update
-	// their entries in their respective caches, along with all of our
-	// unconfirmed transactions and unspent outpoints.
-	n.updateHints(blockHeight)
+// handleConfDetailsAtTip tracks the confirmation height of the txid/output
+// script in order to properly dispatch a confirmation notification after
+// meeting each requests' desired number of confirmations for all current and
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/txnotifier.go]: Fixed.

File: chainntnfs/txnotifier.go [hunk 9]
```diff
@@ -831,31 +833,33 @@ func (n *TxNotifier) ProcessRelevantSpendTx(tx *wire.MsgTx, txHeight int32) erro
...
-		if err := n.updateSpendDetails(prevOut, details); err != nil {
+	// We'll buffer the channel with (2 * number of inputs) to ensure we're
+	// not blocked in the event that the transaction spends registered
+	// outpoints _and_ output scripts,
+	spends := make(chan spend, 2*len(tx.MsgTx().TxIn))
+
+	// We'll set up the onSpend filter callback to stream the fulfilled
+	// spends requests through the channel.
+	onSpend := func(request SpendRequest, details *SpendDetail) {
+		spends <- spend{request, details}
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/txnotifier.go]: Fixed - went with the slice approach.

File: chainntnfs/btcdnotify/btcd_test.go [hunk 10]
```diff
@@ -12,6 +13,20 @@ import (
 	"github.com/lightningnetwork/lnd/channeldb"
 )
 
+var (
+	testScript = []byte{
+		// OP_HASH160
+		0xA9,
+		// OP_DATA_20
+		0x14,
+		// <30-byte hash>
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/btcdnotify/btcd_test.go]: Fixed.

File: chainntnfs/bitcoindnotify/bitcoind_test.go [hunk 11]
```diff
@@ -14,6 +15,20 @@ import (
 	"github.com/lightningnetwork/lnd/channeldb"
 )
 
+var (
+	testScript = []byte{
+		// OP_HASH160
+		0xA9,
+		// OP_DATA_20
+		0x14,
+		// <30-byte hash>
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/bitcoindnotify/bitcoind_test.go]: Fixed.

File: chainntnfs/interface_test.go [hunk 12]
```diff
@@ -1738,9 +1738,9 @@ var blockCatchupTests = []blockCatchupTestCase{
 func TestInterfaces(t *testing.T) {
 	// Initialize the harness around a btcd node which will serve as our
 	// dedicated miner to generate blocks, cause re-orgs, etc. We'll set up
-	// this node with a chain length of 125, so we have plenty of BTC to
-	// play around with.
-	miner, tearDown := chainntnfs.NewMiner(t, nil, true, 25)
+	// this node with a chain length of 500, so that we have plenty of BTC
+	// to play around with and segwit is activated.
+	miner, tearDown := chainntnfs.NewMiner(t, nil, true, 400)
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/interface_test.go]: Fixed.

File: chainntnfs/interface_test.go [hunk 13]
```diff
@@ -1672,40 +1780,43 @@ var ntfnTests = []testCase{
 		test: testMultiClientConfirmationNotification,
 	},
 	{
-		name: "spend ntfn",
-		test: testSpendNotification,
-	},
-	{
-		name: "block epoch",
-		test: testBlockEpochNotification,
+		name: "lazy ntfn consumer",
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/interface_test.go]: Not really, just ordered them in a way that made the most sense to me.

File: chainntnfs/interface_test.go [hunk 14]
```diff
@@ -1815,12 +1927,31 @@ func TestInterfaces(t *testing.T) {
 				notifierType, err)
 		}
 
-		for _, ntfnTest := range ntfnTests {
-			testName := fmt.Sprintf("%v: %v", notifierType,
-				ntfnTest.name)
+		for _, txNtfnTest := range txNtfnTests {
+			testName := fmt.Sprintf("%v %v", notifierType,
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/interface_test.go]: Fixed.

File: lnrpc/chainrpc/config_active.go [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+// Config is the primary configuration struct for the chain notifier RPC server.
+// It contains all the items required for the server to carry out its duties.
+// The fields with struct tags are meant to be parsed as normal configuration
+// options, while if able to be populated, the latter fields MUST also be
+// specified.
+type Config struct {
+	// ChainNotifierMacPath is the path for the chain notifier macaroon. If
+	// unspecified then we assume that the macaroon will be found under the
+	// network directory, named DefaultChainNotifierMacFilename.
+	ChainNotifierMacPath string `long:"chainnotifiermacaroonpath" description:"Path to the chain notifier macaroon"`
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply lnrpc/chainrpc/config_active.go]: Fixed.

File: lnrpc/chainrpc/chainnotifer_server.go [hunk 16]
```diff
@@ -0,0 +1,455 @@
...
+			}
+
+			rpcConfDetails := &ConfDetails{
+				RawTx:       rawTxBuf.Bytes(),
+				BlockHash:   details.BlockHash[:],
+				BlockHeight: details.BlockHeight,
+				TxIndex:     details.TxIndex,
+			}
+
+			conf := &ConfEvent{Event: &ConfEvent_Conf{
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply lnrpc/chainrpc/chainnotifer_server.go]: Fixed.

File: log.go [hunk 17]
```diff
@@ -71,6 +71,7 @@ var (
 	sgnrLog = build.NewSubLogger("SGNR", backendLog.Logger)
 	wlktLog = build.NewSubLogger("WLKT", backendLog.Logger)
 	arpcLog = build.NewSubLogger("ARPC", backendLog.Logger)
+	ntfrLog = build.NewSubLogger("NTFR", backendLog.Logger)
```
Comment (joostjager) [inline log.go]: `ntfrLog` needs to be added to the `subsystemLoggers` list too.

File: lnrpc/chainrpc/chainnotifier.proto [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+    // The index of the confirmed transaction within the transaction.
+    uint32 tx_index = 4;
+}
+
+message Reorg {
+    // TODO(wilmer): need to know how the client will use this first.
+}
+
+message ConfEvent {
+    oneof event {
```
Comment (cfromknecht) [inline lnrpc/chainrpc/chainnotifier.proto]: i've heard mixed reports on using oneof in practice, perhaps better to just say only one will be non-nil? anyone else have insight here?

File: log.go (hunk 17 above)
Comment (wpaulino) [reply log.go]: Fixed.

File: chainntnfs/txnotifier.go [hunk 19]
```diff
@@ -94,6 +108,46 @@ func newSpendNtfnSet() *spendNtfnSet {
...
+// NewConfRequest creates a request for a confirmation notification of either a
+// txid or output script.
+func NewConfRequest(txid *chainhash.Hash, pkScript []byte) (ConfRequest, error) {
+	var r ConfRequest
+	outputScript, err := txscript.ParsePkScript(pkScript)
+	if err != nil {
+		return r, err
+	}
+
+	r.TxID = ZeroHash
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/txnotifier.go]: Fixed.

File: chainntnfs/txnotifier.go [hunk 20]
```diff
@@ -94,6 +108,46 @@ func newSpendNtfnSet() *spendNtfnSet {
...
+	TxID chainhash.Hash
+
+	// PkScript is the public key script of an outpoint created in this
+	// transaction.
+	PkScript txscript.PkScript
+}
+
+// NewConfRequest creates a request for a confirmation notification of either a
+// txid or output script.
+func NewConfRequest(txid *chainhash.Hash, pkScript []byte) (ConfRequest, error) {
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/txnotifier.go]: Fixed.

File: chainntnfs/txnotifier.go [hunk 21]
```diff
@@ -148,6 +150,23 @@ func (r ConfRequest) String() string {
...
 
+// ConfHintKey returns the key that will be used to index the confirmation
+// request's hint within the height hint cache.
+func (r ConfRequest) ConfHintKey() ([]byte, error) {
+	var key []byte
+	if r.TxID == ZeroHash {
+		key = r.PkScript.Script()
+	} else {
+		var txid bytes.Buffer
+		if err := channeldb.WriteElement(&txid, r.TxID); err != nil {
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/txnotifier.go]: Would rather keep it as is to ensure it's using the `channeldb` codec.

File: chainntnfs/btcdnotify/btcd.go [hunk 22]
```diff
@@ -595,24 +611,48 @@ func (b *BtcdNotifier) confDetailsManually(txid *chainhash.Hash, startHeight,
...
+		for txIndex, tx := range block.Transactions {
+			switch {
+			// If the request has a zero hash, then we should match
+			// on the output script instead, so we'll do so.
+			case confRequest.TxID == chainntnfs.ZeroHash:
+				matched := false
+				pkScript := confRequest.PkScript.Script()
+				for _, txOut := range tx.TxOut {
+					if bytes.Equal(txOut.PkScript, pkScript) {
+						matched = true
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/btcdnotify/btcd.go]: Yep, don't know what I was thinking...

File: chainntnfs/neutrinonotify/neutrino.go [hunk 23]
```diff
@@ -517,16 +517,41 @@ func (n *NeutrinoNotifier) historicalConfDetails(targetHash *chainhash.Hash,
...
 			return nil, fmt.Errorf("unable to get block from network: %v", err)
 		}
-		for j, tx := range block.Transactions() {
-			txHash := tx.Hash()
-			if txHash.IsEqual(targetHash) {
-				confDetails := chainntnfs.TxConfirmation{
-					BlockHash:   blockHash,
-					BlockHeight: scanHeight,
-					TxIndex:     uint32(j),
+		for i, tx := range block.Transactions() {
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/neutrinonotify/neutrino.go]: Fixed.

File: chainntnfs/txnotifier.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txnotifier.go]: don't see why channeldb is relevant here tbh, all it does is write the raw, in-order bytes. we haven't used it in the past, we've always compared raw bytes. just seems like extra overhead

[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txnotifier.go]: ahh i see, for the cache keys. makes sense!

**Comment by joostjager** [comment]
Still many commits failing: 
```git rev-list upstream/master..HEAD | xargs -n1 -I{} sh -c 'echo ---- {} ---- && git checkout {} && make'```

**Comment by wpaulino** [comment]
Rebased.

**Pull request merged**